code/game/machinery/computer/robot.dm

```diff
@@ -132,3 +165,26 @@
 					s.start()
 					D.visible_message("<span class='danger'>\the [D] self-destructs!</span>")
 					D.gib()
+		if("extract")
+			if(!GLOB.upload_code)
+				GLOB.upload_code = random_nukecode()
+
+			message_admins("[ADMIN_LOOKUPFLW(usr)] is extracting the upload key!")
+			extracting = TRUE
+			ui_update()
+			if(allowed(usr))
+				say("Credentials successfully verified, commencing extraction.")
+				sleep(120)
+
+			else
+				var/message = "ALERT: UNAUTHORIZED UPLOAD KEY EXTRACTION AT [get_area_name(loc, TRUE)]"
+				radio.talk_into(src, message, radio_channel)
+				sleep(600)
+				say("Extraction at 50%")
+				sleep(600)
```

Comment:
Can you do this withou tthe sleep? Especially sinces its in ui_act

Reply:
I'm not entirely certain how to go about that

Reply:
Would a Callback be better?

Reply:
I'll just move it into a seperate proc that isn't in ui_act